server/schedule/operator/builder_test.go

```diff
@@ -0,0 +1,256 @@
+	s.cluster = mockcluster.NewCluster(opts)
+	s.cluster.AddLabelsStore(1, 0, map[string]string{"zone": "z1", "host": "h1"})
+	s.cluster.AddLabelsStore(2, 0, map[string]string{"zone": "z1", "host": "h1"})
+	s.cluster.AddLabelsStore(3, 0, map[string]string{"zone": "z1", "host": "h1"})
+	s.cluster.AddLabelsStore(4, 0, map[string]string{"zone": "z1", "host": "h1"})
+	s.cluster.AddLabelsStore(5, 0, map[string]string{"zone": "z1", "host": "h1"})
+	s.cluster.AddLabelsStore(6, 0, map[string]string{"zone": "z1", "host": "h2"})
+	s.cluster.AddLabelsStore(7, 0, map[string]string{"zone": "z1", "host": "h2"})
+	s.cluster.AddLabelsStore(8, 0, map[string]string{"zone": "z2", "host": "h1"})
+	s.cluster.AddLabelsStore(9, 0, map[string]string{"zone": "z2", "host": "h2"})
+	s.cluster.AddLabelsStore(10, 0, map[string]string{"zone": "z3", "host": "h1", "noleader": "true"})
+}
+
+func (s *testBuilderSuite) TestNewBuilder(c *C) {
+	peers := []*metapb.Peer{{Id: 11, StoreId: 1}, {Id: 12, StoreId: 2, IsLearner: true}}
+	region := core.NewRegionInfo(&metapb.Region{Id: 42, Peers: peers}, peers[0])
+	builder := NewBuilder("test", s.cluster, region)
+	c.Assert(builder.err, IsNil)
+	c.Assert(builder.originPeers.Len(), Equals, 2)
+	c.Assert(builder.originPeers.Get(1), DeepEquals, peers[0])
+	c.Assert(builder.originPeers.Get(2), DeepEquals, peers[1])
+	c.Assert(builder.originLeader, Equals, uint64(1))
+	c.Assert(builder.targetPeers.Len(), Equals, 2)
+	c.Assert(builder.targetPeers.Get(1), DeepEquals, peers[0])
+	c.Assert(builder.targetPeers.Get(2), DeepEquals, peers[1])
+	region = region.Clone(core.WithLeader(nil))
+	builder = NewBuilder("test", nil, region)
+	c.Assert(builder.err, NotNil)
+}
+
+func (s *testBuilderSuite) newBuilder() *Builder {
+	peers := []*metapb.Peer{
+		{Id: 11, StoreId: 1},
+		{Id: 12, StoreId: 2},
+		{Id: 13, StoreId: 3, IsLearner: true},
+	}
+	region := core.NewRegionInfo(&metapb.Region{Id: 1, Peers: peers}, peers[0])
+	return NewBuilder("test", s.cluster, region)
+}
+
+func (s *testBuilderSuite) TestRecord(c *C) {
+	c.Assert(s.newBuilder().AddPeer(&metapb.Peer{StoreId: 1}).err, NotNil)
+	c.Assert(s.newBuilder().AddPeer(&metapb.Peer{StoreId: 4}).err, IsNil)
+	c.Assert(s.newBuilder().PromoteLearner(1).err, NotNil)
+	c.Assert(s.newBuilder().PromoteLearner(3).err, IsNil)
+	c.Assert(s.newBuilder().SetLeader(1).SetLeader(2).err, IsNil)
+	c.Assert(s.newBuilder().SetLeader(3).err, NotNil)
+	c.Assert(s.newBuilder().RemovePeer(4).err, NotNil)
+	c.Assert(s.newBuilder().AddPeer(&metapb.Peer{StoreId: 4, IsLearner: true}).RemovePeer(4).err, IsNil)
+	c.Assert(s.newBuilder().SetLeader(2).RemovePeer(2).err, NotNil)
+	c.Assert(s.newBuilder().PromoteLearner(4).err, NotNil)
+	c.Assert(s.newBuilder().SetLeader(4).err, NotNil)
+
+	m := map[uint64]*metapb.Peer{
+		2: {StoreId: 2},
+		3: {StoreId: 3, IsLearner: true},
+		4: {StoreId: 4},
+	}
+	builder := s.newBuilder().SetPeers(m).SetLightWeight()
```

Comment:
What will happen if `m` is invalid? For example:
``` go
map[uint64]*metapb.Peer{
    1: {StoreId: 2},
    2: {StoreId: 0},
    0: {StoreId: 3},
}
```

Reply:
Always use the id in metapb.Peer. Using map here is for lagacy code more easy to use.

Reply:
I prefer to do some check in `SetPeers()`.

Reply:
em, I've just added some description about the behaviour, PTAL.

Reply:
It's using map key to check whether current leader is in new peers.

Reply:
And I mainly worry about 0 store id.

Reply:
Make sense. I'll change to check mismatch and return error.